egs/librispeech/asr/simple_v1/bpe_ctc_att_conformer_train.py

```diff
@@ -0,0 +1,633 @@
+        default=0.0,
+        help='weight decay (L2 penalty) for Noam optimizer.'
+    )
+    parser.add_argument(
+        '--accum-grad',
+        type=int,
+        default=1,
+        help="Number of gradient accumulation.")
+    parser.add_argument(
+        '--att-rate',
+        type=float,
+        # Reference: https://zenodo.org/record/4604066#.YNPzOOgzaUk
+        # as ctc_weight=0.1 in previous reference.
+        default=0.7,
+        help="Attention loss rate.")
+    parser.add_argument(
+        '--nhead',
+        type=int,
+        default=8,
+        help="Number of attention heads in transformer.")
+    parser.add_argument(
+        '--attention-dim',
+        type=int,
+        default=512,
+        help="Number of units in transformer attention layers.")
+    parser.add_argument(
+        '--is-espnet-structure',
+        type=str2bool,
+        default=True,
+        help='When enabled, the conformer will have the ' \
+             'same structure like espnet')
+    parser.add_argument(
+        '--vgg-frontend',
+        type=str2bool,
+        default=False,
+        help='When enabled, it uses vgg style network for subsampling')
+    parser.add_argument(
+        '--espnet-identical-model',
+        type=str2bool,
+        default=False,
+        help='When enabled, train an identical model to the espnet SOTA released model'
+        "url: https://zenodo.org/record/4604066#.YNAAHmgzZPY")
+    parser.add_argument(
+        '--tensorboard',
+        type=str2bool,
+        default=True,
+        help='Should various information be logged in tensorboard.'
+    )
+    return parser
+
+
+def run(rank, world_size, args):
+    '''
+    Args:
+      rank:
+        It is a value between 0 and `world_size-1`, which is
+        passed automatically by `mp.spawn()` in :func:`main`.
+        The node with rank 0 is responsible for saving checkpoint.
+      world_size:
+        Number of GPUs for DDP training.
+      args:
+        The return value of get_parser().parse_args()
+    '''
+    assert args.start_epoch >= 0
+    model_type = args.model_type
+    curr_epoch = args.start_epoch + 1
+    num_epochs = args.num_epochs
+    accum_grad = args.accum_grad
+    att_rate = args.att_rate
+    attention_dim = args.attention_dim
+    nhead=args.nhead
+
+    fix_random_seed(42)
+    setup_dist(rank, world_size, args.master_port)
+
+    exp_dir = Path(f'exp-duration-200-feat_batchnorm-bpe-lrfactor{args.lr_factor}-{model_type}-{attention_dim}-{nhead}-noam')
+    setup_logger(f'{exp_dir}/log/log-train-{rank}')
+    if args.tensorboard and rank == 0:
+        tb_writer = SummaryWriter(log_dir=f'{exp_dir}/tensorboard')
+    else:
+        tb_writer = None
+
+    device_id = rank
+    device = torch.device('cuda', device_id)
+
+
+
+    if not torch.cuda.is_available():
+        logging.error('no gpu detected!')
+        sys.exit(-1)
+
+    # TODO(Liyong Guo) make this configurable.
+    lang_dir = Path('data/en_token_list/bpe_unigram5000/')
```

Comment:
How do you generate the directory `data/en_token_list/bpe_unigram5000/`?
I don't find any code responsible for it.

Reply:
Will the code to generate bpe related files.
I will make a pr to your branch. @glynpu

Reply:
Currently, they are downloaded from the [model zoo](https://huggingface.co/GuoLiyong/snowfall_bpe_model/tree/main/data/en_token_list/bpe_unigram5000).

Please refer to the bpe_run.sh, which contains following downloading code:
```
    git clone https://huggingface.co/GuoLiyong/snowfall_bpe_model
    for sub_dir in data; do
      ln -sf snowfall_bpe_model/$sub_dir ./
    done
```

Actually, I deliberately don't sumit the code about training bpe model, because this pr is mainly about training pipeline.